File: geotrellis-python/geotrellis/spark/io/avro/codecs/TileCodecs.py
from __future__ import absolute_import
from geotrellis.spark.io.avro.AvroRecordCodec import AvroRecordCodec
from geotrellis.python.edited_avro.avro_builder import AvroSchemaBuilder
from geotrellis.raster.CellType import (
        ShortConstantNoDataCellType, ShortCellType, ShortUserDefinedNoDataCellType,
        UShortConstantNoDataCellType, UShortCellType, UShortUserDefinedNoDataCellType,
        IntConstantNoDataCellType, IntCellType, IntUserDefinedNoDataCellType,
        FloatConstantNoDataCellType, FloatCellType, FloatUserDefinedNoDataCellType,
        DoubleConstantNoDataCellType, DoubleCellType, DoubleUserDefinedNoDataCellType,
        ByteConstantNoDataCellType, ByteCellType, ByteUserDefinedNoDataCellType,
        UByteConstantNoDataCellType, UByteCellType, UByteUserDefinedNoDataCellType
        )
from geotrellis.raster.ShortArrayTile import ShortArrayTile
from geotrellis.raster.UShortArrayTile import UShortArrayTile
from geotrellis.raster.IntArrayTile import IntArrayTile
from geotrellis.raster.FloatArrayTile import FloatArrayTile
from geotrellis.raster.DoubleArrayTile import DoubleArrayTile
from geotrellis.raster.ByteArrayTile import ByteArrayTile
from geotrellis.raster.UByteArrayTile import UByteArrayTile
from geotrellis.raster.BitArrayTile import BitArrayTile
from geotrellis.raster.MultibandTile import MultibandTile
from geotrellis.raster.ArrayMultibandTile import ArrayMultibandTile

from geotrellis.spark.io.avro.codecs.Implicits import tileUnionCodec

from geotrellis.raster.package_scala import SHORTMIN, BYTEMIN, isNoData, shortNODATA, ushortNODATA, NODATA, byteNODATA, ubyteNODATA

import avro.schema

import array

class ShortArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, ShortArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("ShortArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells",
                _.begin_array(_.create_int()).end())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_int(),
                    _.create_null()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist()
        if tile.cellType == ShortConstantNoDataCellType:
            dct["noDataValue"] = shortNODATA
        elif tile.cellType == ShortCellType:
            dct["noDataValue"] = None # TODO find matching java's null alternative
        elif isinstance(tile.cellType, ShortUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        arr = array.array('h', dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is None:
            cellType = ShortCellType
        elif nodata == SHORTMIN:
            cellType = ShortConstantNoDataCellType
        else:
            cellType = ShortUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return ShortArrayTile.applyStatic(arr, cols, rows, cellType)

class UShortArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, UShortArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("UShortArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells",
                _.begin_array(_.create_int()).end())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_int(),
                    _.create_null()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist()
        if tile.cellType == UShortConstantNoDataCellType:
            dct["noDataValue"] = ushortNODATA
        elif tile.cellType == UShortCellType:
            dct["noDataValue"] = None
        elif isinstance(tile.cellType, UShortUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        arr = array.array('H', dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is None:
            cellType = UShortCellType
        elif nodata == 0:
            cellType = UShortConstantNoDataCellType
        else:
            cellType = UShortUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return UShortArrayTile.applyStatic(arr, cols, rows, cellType)

class IntArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, IntArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("IntArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells",
                _.begin_array(_.create_int()).end())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_int(),
                    _.create_null()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist()
        if tile.cellType == IntConstantNoDataCellType:
            dct["noDataValue"] = NODATA
        elif tile.cellType == IntCellType:
            dct["noDataValue"] = None
        elif isinstance(tile.cellType, IntUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        #arr = array.array('i', dct["cells"])
        # <debug>
        cells = dct["cells"]
        print("dct.cells: {c}".format(c=cells))
        arr = array.array('i', cells)
        #print("array.array: {arr}".format(arr=arr))
        #print("arr.arr.lst: {arr}".format(arr=arr.tolist()))
        # </debug>
        nodata = dct["noDataValue"]
        if nodata is None:
            cellType = IntCellType
        elif isNoData(nodata):
            cellType = IntConstantNoDataCellType
        else:
            cellType = IntUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return IntArrayTile.applyStatic(arr, cols, rows, cellType)

class FloatArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, FloatArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("FloatArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells",
                _.begin_array(_.create_float()).end())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_float(),
                    _.create_boolean()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist()
        if tile.cellType == FloatConstantNoDataCellType:
            dct["noDataValue"] = True
        elif tile.cellType == FloatCellType:
            dct["noDataValue"] = False
        elif isinstance(tile.cellType, FloatUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected. {tp}".format(ct = tile.cellType, tp = type(tile.cellType)))

    def decode(self, dct):
        arr = array.array('f', dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is True:
            cellType = FloatConstantNoDataCellType
        elif nodata is False:
            cellType = FloatCellType
        else:
            cellType = FloatUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return FloatArrayTile.applyStatic(arr, cols, rows, cellType)

class DoubleArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, DoubleArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("DoubleArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells",
                _.begin_array(_.create_double()).end())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_double(),
                    _.create_boolean()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist()
        if tile.cellType == DoubleConstantNoDataCellType:
            dct["noDataValue"] = True
        elif tile.cellType == DoubleCellType:
            dct["noDataValue"] = False
        elif isinstance(tile.cellType, DoubleUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        arr = array.array('d', dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is True:
            cellType = DoubleConstantNoDataCellType
        elif nodata is False:
            cellType = DoubleCellType
        else:
            cellType = DoubleUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return DoubleArrayTile.applyStatic(arr, cols, rows, cellType)

class ByteArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, ByteArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("ByteArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells", _.create_bytes())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_int(),
                    _.create_null()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        #dct["cells"] = tile.array.tolist() # TODO wrap if needed
        dct["cells"] = tile.array.tostring()
        if tile.cellType == ByteConstantNoDataCellType:
            dct["noDataValue"] = byteNODATA
        elif tile.cellType == ByteCellType:
            dct["noDataValue"] = None
        elif isinstance(tile.cellType, ByteUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        arr = array.array('b')
        arr.fromstring(dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is None:
            cellType = ByteCellType
        elif nodata == BYTEMIN:
            cellType = ByteConstantNoDataCellType
        else:
            cellType = ByteUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return ByteArrayTile.applyStatic(arr, cols, rows, cellType)

class UByteArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, UByteArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("UByteArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells", _.create_bytes())
        _.add_field("noDataValue",
                _.begin_union(
                    _.create_int(),
                    _.create_null()
                    ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        #dct["cells"] = tile.array.tolist() # TODO wrap if needed
        dct["cells"] = tile.array.tostring()
        if tile.cellType == UByteConstantNoDataCellType:
            dct["noDataValue"] = ubyteNODATA
        elif tile.cellType == UByteCellType:
            dct["noDataValue"] = None
        elif isinstance(tile.cellType, UByteUserDefinedNoDataCellType):
            dct["noDataValue"] = tile.cellType.noDataValue
        else:
            raise Exception(
                    "CellType {ct} was unexpected".format(ct = tile.cellType))

    def decode(self, dct):
        arr = array.array('B')
        arr.fromstring(dct["cells"])
        nodata = dct["noDataValue"]
        if nodata is None:
            cellType = UByteCellType
        elif nodata == 0:
            cellType = UByteConstantNoDataCellType
        else:
            cellType = UByteUserDefinedNoDataCellType(nodata)

        cols = dct["cols"]
        rows = dct["rows"]
        return UByteArrayTile.applyStatic(arr, cols, rows, cellType)

class BitArrayTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, BitArrayTile)

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("BitArrayTile", namespace = "geotrellis.raster")
        _.add_field("cols", _.create_int())
        _.add_field("rows", _.create_int())
        _.add_field("cells", _.create_bytes())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        dct["cols"] = tile.cols
        dct["rows"] = tile.rows
        dct["cells"] = tile.array.tolist() # TODO wrap if needed

    def decode(self, dct):
        arr = array.array('b', dct["cells"])
        cols = dct["cols"]
        rows = dct["rows"]
        return BitArrayTile.applyStatic(arr, cols, rows)

class MultibandTileCodec(AvroRecordCodec):
    def __init__(self):
        AvroRecordCodec.__init__(self, MultibandTile)
        self._tileUnionCodec = tileUnionCodec()

    @property
    def schema(self):
        _ = AvroSchemaBuilder()
        _.begin_record("ArrayMultibandTile", namespace = "geotrellis.raster")
        _.add_field("bands", _.begin_array(
            #_.begin_with_schema_json(self._tileUnionCodec.schema.to_json())
            self._tileUnionCodec.schema.to_json()
            ).end())
        dct = _.end()
        return avro.schema.make_avsc_object(dct)

    def _encode(self, tile, dct):
        bands = [tile.band(i) for i in xrange(0, tile.bandCount)]
        dct["bands"] = map(self._tileUnionCodec.encode, bands)

    def decode(self, dct):
        bands = map(self._tileUnionCodec.decode, dct["bands"])
        return ArrayMultibandTile(bands)

File: layers/src/main/scala/geotrellis/layers/avro/codecs/TileCodecs.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.layers.avro.codecs

import java.nio.ByteBuffer

import geotrellis.raster._
import geotrellis.layers.avro._
import geotrellis.layers.avro.codecs.Implicits._

import org.apache.avro.SchemaBuilder
import org.apache.avro.generic._

import scala.collection.JavaConverters._
import scala.util.Try

trait TileCodecs {
  implicit def shortArrayTileCodec: AvroRecordCodec[ShortArrayTile] = new AvroRecordCodec[ShortArrayTile] {
    def schema = SchemaBuilder
      .record("ShortArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().array().items().intType().noDefault()
      .name("noDataValue").`type`().unionOf().intType().and().nullType().endUnion().intDefault(Short.MinValue.toInt)
      .endRecord()

    def encode(tile: ShortArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      // _* expansion is important, otherwise we get List[Array[Short]] instead of List[Short]
      rec.put("cells", java.util.Arrays.asList(tile.array:_*))
      tile.cellType match {
        case ShortConstantNoDataCellType => rec.put("noDataValue", Short.MinValue.toInt)
        case ShortUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd.toInt)
        case ShortCellType => rec.put("noDataValue", null)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[java.util.Collection[Int]]
        .asScala // notice that Avro does not have native support for Short primitive
        .map(_.toShort)
        .toArray
      val cellType = Option(rec[Int]("noDataValue")) match {
        case Some(nd) if nd == Short.MinValue.toInt => ShortConstantNoDataCellType
        case Some(nd) => ShortUserDefinedNoDataCellType(nd.toShort)
        case None => ShortCellType
      }

      ShortArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  implicit def uShortArrayTileCodec: AvroRecordCodec[UShortArrayTile] = new AvroRecordCodec[UShortArrayTile] {
    def schema = SchemaBuilder
      .record("UShortArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().array().items().intType().noDefault()
      .name("noDataValue").`type`().unionOf().intType().and().nullType().endUnion().intDefault(0)
      .endRecord()

    def encode(tile: UShortArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      // _* expansion is important, otherwise we get List[Array[Short]] instead of List[Short]
      rec.put("cells", java.util.Arrays.asList(tile.array:_*))
      tile.cellType match {
        case UShortConstantNoDataCellType => rec.put("noDataValue", 0)
        case UShortUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd.toInt)
        case UShortCellType => rec.put("noDataValue", null)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[java.util.Collection[Int]]
        .asScala // notice that Avro does not have native support for Short primitive
        .map(_.toShort)
        .toArray
      val cellType = Option(rec[Int]("noDataValue")) match {
        case Some(nd) if nd == 0 => UShortConstantNoDataCellType
        case Some(nd) => UShortUserDefinedNoDataCellType(nd.toShort)
        case None => UShortCellType
      }
      UShortArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  implicit def intArrayTileCodec: AvroRecordCodec[IntArrayTile] = new AvroRecordCodec[IntArrayTile] {
    def schema = SchemaBuilder
      .record("IntArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().array().items().intType().noDefault()
      .name("noDataValue").`type`().unionOf().intType().and().nullType().endUnion().intDefault(Int.MinValue)
      .endRecord()

    def encode(tile: IntArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", java.util.Arrays.asList(tile.array:_*))
      tile.cellType match {
        case IntConstantNoDataCellType => rec.put("noDataValue", NODATA)
        case IntUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd)
        case IntCellType => rec.put("noDataValue", null)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[java.util.Collection[Int]]
        .asScala
        .toArray[Int]
      val cellType = Option(rec[Int]("noDataValue")) match {
        case Some(nd) if isNoData(nd) => IntConstantNoDataCellType
        case Some(nd) => IntUserDefinedNoDataCellType(nd)
        case None => IntCellType
      }
      IntArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  /** Avro serialization doesn't support Float.NaN or Double.NaN. Whereas a
    * union of number and null is sufficient in cases where the nodata value
    * for some domain is can be serialized (Int.MinValue is just another
    * integer and, therefore, serializable without difficulty), we are in need
    * of an alternative strategy for floating point serialization.
    *
    * To this end, we've serialized with a union of boolean and floating point
    * values.
    * noDataValue can either be:
    * 1. true (and, therefore, ConstantNoData)
    * 2. false (NoNoData)
    * - OR -
    * 3. a floating point value (which is a UserDefinedNoDataValue's value)
    */
  implicit def floatArrayTileCodec: AvroRecordCodec[FloatArrayTile] = new AvroRecordCodec[FloatArrayTile] {
    def schema = SchemaBuilder
      .record("FloatArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().array().items().floatType().noDefault()
      .name("noDataValue").`type`().unionOf().booleanType().and().floatType().endUnion().booleanDefault(true)
      .endRecord()

    def encode(tile: FloatArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", java.util.Arrays.asList(tile.array:_*))
      tile.cellType match {
        case FloatConstantNoDataCellType => rec.put("noDataValue", true)
        case FloatUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd)
        case FloatCellType => rec.put("noDataValue", false)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[java.util.Collection[Float]]
        .asScala
        .toArray[Float]
      val cellType =
        Try {
          if (rec[Boolean]("noDataValue") == true) FloatConstantNoDataCellType
          else FloatCellType
        } getOrElse {
          FloatUserDefinedNoDataCellType(rec[Float]("noDataValue"))
        }
      FloatArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  /** Avro serialization doesn't support Float.NaN or Double.NaN. Whereas a
    * union of number and null is sufficient in cases where the nodata value
    * for some domain is can be serialized (Int.MinValue is just another
    * integer and, therefore, serializable without difficulty), we are in need
    * of an alternative strategy for floating point serialization.
    *
    * To this end, we've serialized with a union of boolean and floating point
    * values.
    * noDataValue can either be:
    * 1. true (and, therefore, ConstantNoData)
    * 2. false (NoNoData)
    * - OR -
    * 3. a floating point value (which is a UserDefinedNoDataValue's value)
    */
  implicit def doubleArrayTileCodec: AvroRecordCodec[DoubleArrayTile] = new AvroRecordCodec[DoubleArrayTile] {
    def schema = SchemaBuilder
      .record("DoubleArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().array().items().doubleType().noDefault()
      .name("noDataValue").`type`().unionOf().booleanType().and().doubleType().endUnion().booleanDefault(true)
      .endRecord()

    def encode(tile: DoubleArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", java.util.Arrays.asList(tile.array:_*))
      tile.cellType match {
        case DoubleConstantNoDataCellType => rec.put("noDataValue", true)
        case DoubleUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd)
        case DoubleCellType => rec.put("noDataValue", false)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[java.util.Collection[Double]]
        .asScala
        .toArray[Double]
      val cellType =
        Try {
          if (rec[Boolean]("noDataValue") == true) DoubleConstantNoDataCellType
          else DoubleCellType
        } getOrElse {
          DoubleUserDefinedNoDataCellType(rec[Double]("noDataValue"))
        }
      DoubleArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  implicit def byteArrayTileCodec: AvroRecordCodec[ByteArrayTile] = new AvroRecordCodec[ByteArrayTile] {
    def schema = SchemaBuilder
      .record("ByteArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().bytesType().noDefault()
      .name("noDataValue").`type`().unionOf().intType().and().nullType().endUnion().intDefault(Byte.MinValue.toInt)
      .endRecord()

    def encode(tile: ByteArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", ByteBuffer.wrap(tile.array))
      tile.cellType match {
        case ByteConstantNoDataCellType => rec.put("noDataValue", byteNODATA.toInt)
        case ByteUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd.toInt)
        case ByteCellType => rec.put("noDataValue", null)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[ByteBuffer]
        .array()
      val cellType = Option(rec[Int]("noDataValue")) match {
        case Some(nd) if nd == Byte.MinValue.toInt => ByteConstantNoDataCellType
        case Some(nd) => ByteUserDefinedNoDataCellType(nd.toByte)
        case None => ByteCellType
      }
      ByteArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  implicit def uByteArrayTileCodec: AvroRecordCodec[UByteArrayTile] = new AvroRecordCodec[UByteArrayTile] {
    def schema = SchemaBuilder
      .record("UByteArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().bytesType().noDefault()
      .name("noDataValue").`type`().unionOf().intType().and().nullType().endUnion().intDefault(0)
      .endRecord()

    def encode(tile: UByteArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", ByteBuffer.wrap(tile.array))
      tile.cellType match {
        case UByteConstantNoDataCellType => rec.put("noDataValue", 0)
        case UByteUserDefinedNoDataCellType(nd) => rec.put("noDataValue", nd.toInt)
        case UByteCellType => rec.put("noDataValue", null)
        case _ => sys.error(s"Cell type ${tile.cellType} was unexpected")
      }
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells")
        .asInstanceOf[ByteBuffer]
        .array()
      val cellType = Option(rec[Int]("noDataValue")) match {
        case Some(nd) if nd == 0 => UByteConstantNoDataCellType
        case Some(nd) => UByteUserDefinedNoDataCellType(nd.toByte)
        case None => UByteCellType
      }
      UByteArrayTile(array, rec[Int]("cols"), rec[Int]("rows"), cellType)
    }
  }

  implicit def bitArrayTileCodec: AvroRecordCodec[BitArrayTile] = new AvroRecordCodec[BitArrayTile] {
    def schema = SchemaBuilder
      .record("BitArrayTile").namespace("geotrellis.raster")
      .fields()
      .name("cols").`type`().intType().noDefault()
      .name("rows").`type`().intType().noDefault()
      .name("cells").`type`().bytesType().noDefault()
      .endRecord()

    def encode(tile: BitArrayTile, rec: GenericRecord) = {
      rec.put("cols", tile.cols)
      rec.put("rows", tile.rows)
      rec.put("cells", ByteBuffer.wrap(tile.array))
    }

    def decode(rec: GenericRecord) = {
      val array  = rec.get("cells").asInstanceOf[ByteBuffer].array()
      BitArrayTile(array, rec[Int]("cols"), rec[Int]("rows"))
    }
  }

  implicit def multibandTileCodec: AvroRecordCodec[MultibandTile] = new AvroRecordCodec[MultibandTile] {
    def schema = SchemaBuilder
      .record("ArrayMultibandTile").namespace("geotrellis.raster")
      .fields()
      .name("bands").`type`().array().items.`type`(tileUnionCodec.schema).noDefault()
      .endRecord()

    def encode(tile: MultibandTile, rec: GenericRecord) = {
      val bands = for (i <- 0 until tile.bandCount) yield tile.band(i)
      rec.put("bands", bands.map(tileUnionCodec.encode).asJavaCollection)
    }

    def decode(rec: GenericRecord) = {
      val bands = rec.get("bands")
        .asInstanceOf[java.util.Collection[GenericRecord]]
        .asScala // notice that Avro does not have native support for Short primitive
        .map(tileUnionCodec.decode)
        .toArray

      ArrayMultibandTile(bands)
    }
  }
}

object TileCodecs extends TileCodecs
